staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -893,3 +895,209 @@ func (m *mockTxnStore) Transaction(txns ...Transaction) *TransactionError {
+
+	// Run replace
+	replaceFifo(t, fifo, []interface{}{pAdd, pKeep, pModUpdated})
+
+	err = m.waitForEventCount(ctx, 4, 5*time.Second)
+	if err != nil {
+		t.Fatalf("Controller failed to receive all events, got: %v", m.getHistory())
+	}
+
+	expectItems := []any{
+		pAdd,
+		pKeep,
+		pModUpdated,
+	}
+	items := sortItemsByKey(store.List(), DeletionHandlingMetaNamespaceKeyFunc)
+	if !apiequality.Semantic.DeepEqual(expectItems, items) {
+		t.Error(cmp.Diff(expectItems, items))
+	}
+	expectHistory := []eventRecord{
+		{Action: "add", Key: "default/pod-add"},
+		{Action: "delete", Key: "default/pod-del"},
+		{Action: "update", Key: "default/pod-keep"},
+		{Action: "update", Key: "default/pod-mod"},
+	}
+	if !apiequality.Semantic.DeepEqual(expectHistory, m.getHistory()) {
+		t.Error(cmp.Diff(expectHistory, m.getHistory()))
+	}
+}
+
+type eventRecord struct {
+	Action string
+	Key    string
+}
+
+type eventRecorder struct {
+	historyLock sync.Mutex
+	history     []eventRecord
+
+	updateCh chan bool
+}
+
+func (m *eventRecorder) OnAdd(obj interface{}, _ bool) {
+	m.record("add", obj)
+}
+func (m *eventRecorder) OnUpdate(_, obj interface{}) {
+	m.record("update", obj)
+}
+func (m *eventRecorder) OnDelete(obj interface{}) {
+	m.record("delete", obj)
+}
+
+func (m *eventRecorder) record(action string, obj interface{}) {
+	m.historyLock.Lock()
+	defer m.historyLock.Unlock()
+	key, _ := DeletionHandlingMetaNamespaceKeyFunc(obj)
+	m.history = append(m.history, eventRecord{Action: action, Key: key})
+
+	select {
+	case m.updateCh <- true:
+	default:
+	}
+}
+
+func (m *eventRecorder) clearHistory() {
+	m.historyLock.Lock()
+	defer m.historyLock.Unlock()
+	m.history = []eventRecord{}
+}
+
+func (m *eventRecorder) getHistory() []eventRecord {
+	m.historyLock.Lock()
+	historyCopy := make([]eventRecord, len(m.history))
+	copy(historyCopy, m.history)
+	m.historyLock.Unlock()
+
+	sortEvents := func(events []eventRecord) {
+		sort.Slice(events, func(i, j int) bool {
+			if events[i].Key != events[j].Key {
+				return events[i].Key < events[j].Key
+			}
+			return events[i].Action < events[j].Action
+		})
+	}
+	sortEvents(historyCopy)
+
+	return historyCopy
+}
+
+func (m *eventRecorder) waitForEventCount(ctx context.Context, count int, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	for {
+		m.historyLock.Lock()
+		currentCount := len(m.history)
+		m.historyLock.Unlock()
+
+		if currentCount >= count {
+			return nil
+		}
+
+		select {
+		case <-m.updateCh:
+			continue
+		case <-ctx.Done():
+			return fmt.Errorf("context canceled waiting for %d events, currently have %d", count, currentCount)
+		}
+	}
+}
+
+func sortItemsByKey(items []interface{}, keyFunc KeyFunc) []interface{} {
+	sort.Slice(items, func(i, j int) bool {
+		keyI, _ := keyFunc(items[i])
+		keyJ, _ := keyFunc(items[j])
+		return keyI < keyJ
+	})
+	return items
+}
+
+func addFifo(t *testing.T, fifo Queue, obj interface{}) {
```

Comment:
What would happen if every tests started adding such methods just to shorten `if err != nil`. We would get more boilerplate then real code. If you want to do that I would recomend looking into using testify
"github.com/stretchr/testify/require"

Like
```
require.NoError(t, fifo.Add(obj), "failed to add")
```